crates/wasmparser/src/validator/func.rs

```diff
@@ -311,9 +313,10 @@ mod tests {
 macro_rules! define_visit_operator {
     ($(@$proposal:ident $op:ident $({ $($arg:ident: $argty:ty),* })? => $visit:ident)*) => {
         $(
-            fn $visit(&mut self, offset: usize $($(,$arg: $argty)*)?) -> Result<()> {
-                self.validator.with_resources(&self.resources)
-                    .$visit(offset $($(,$arg)*)?)
+            fn $visit(&mut self $($(,$arg: $argty)*)?) -> Result<()> {
+                // TODO: whatever happens if the caller does not set `op_offset` ahead of time?
```

Comment:
The public `VisitOperator` implementation on this type is problematic. The validation today depends on offsets always getting set-up correctly for more than just the error messages. In particular it sets `self.end_which_emptied_control` to the offset of the very last `end` for the function, which is then checked in the `finish` method.

On the other hand, it is possible for the users to call `reader.visit_operator(func_validator)` while forgetting to set-up the `op_offset`. This is distinct from `FuncValidator::op` which does the right thing.

I would say resolving this will require to remove or move this implementation to some other type which forces callers to set up the offset one way or another, much like how `with_resources` below does.

Maybe something like

```rust
fn visit_operator(&self, reader: BinaryReader) -> ... { }
```

Reply:
Hm yeah that is indeed unfortunate. It's fine to implement the `end_which_emptied_control` bit entirely differently, that was just the first thing I reached for.

One possibility, though, could perhaps be a default-does-nothing trait method which is along the lines of "I'm about to start visiting this offset", which could be called at the start of `visit_operator` and ignored by almost everything except the validator?

Reply:
Yeah, adding a `visit_offset` or such to the `VisitOperator` is an option. It is an obvious way forward, but it isn’t clear to me if it is the _best_ way forward quite yet.

I want to give it some thought and maybe experiment with them to see how things work out if done differently.

Reply:
I ended up with what I think is a simpler approach for now of adding a `visitor` method to the `FuncValidator` which returns an opaque `impl VisitOperator`.